cli/mclass_info.c

```diff
@@ -0,0 +1,249 @@
+        goto out;
+    }
+
+    yaml_scalar_event_initialize(
+        &event,
+        NULL,
+        (yaml_char_t *)YAML_STR_TAG,
+        (yaml_char_t *)PATH_KEY,
+        sizeof(PATH_KEY) - 1,
+        1,
+        0,
+        YAML_PLAIN_SCALAR_STYLE);
+    if (!yaml_emitter_emit(&emitter, &event)) {
+        rc = EIO;
+        goto out;
+    }
+
+    yaml_scalar_event_initialize(
+        &event,
+        NULL,
+        (yaml_char_t *)YAML_STR_TAG,
+        (yaml_char_t *)info.mi_path,
+        strnlen(info.mi_path, sizeof(info.mi_path)),
+        1,
+        0,
+        YAML_PLAIN_SCALAR_STYLE);
+    if (!yaml_emitter_emit(&emitter, &event)) {
+        rc = EIO;
+        goto out;
+    }
+
+    yaml_scalar_event_initialize(
+        &event,
+        NULL,
+        (yaml_char_t *)YAML_STR_TAG,
+        (yaml_char_t *)ALLOCATED_BYTES_KEY,
+        sizeof(ALLOCATED_BYTES_KEY) - 1,
+        1,
+        0,
+        YAML_PLAIN_SCALAR_STYLE);
+    if (!yaml_emitter_emit(&emitter, &event)) {
+        rc = EIO;
+        goto out;
+    }
+
+    rc = snprintf(buf, sizeof(buf), "%lu", info.mi_allocated_bytes);
+    if (rc < 0) {
+        return EBADMSG;
+    } else if (rc >= sizeof(buf)) {
+        return EMSGSIZE;
+    }
+
+    yaml_scalar_event_initialize(
+        &event,
+        NULL,
+        (yaml_char_t *)YAML_STR_TAG,
+        (yaml_char_t *)buf,
+        strnlen(buf, sizeof(buf)),
+        1,
+        0,
+        YAML_PLAIN_SCALAR_STYLE);
+    if (!yaml_emitter_emit(&emitter, &event)) {
+        rc = EIO;
+        goto out;
+    }
+
+    yaml_scalar_event_initialize(
+        &event,
+        NULL,
+        (yaml_char_t *)YAML_STR_TAG,
+        (yaml_char_t *)USED_BYTES_KEY,
+        sizeof(USED_BYTES_KEY) - 1,
+        1,
+        0,
+        YAML_PLAIN_SCALAR_STYLE);
+    if (!yaml_emitter_emit(&emitter, &event)) {
+        rc = EIO;
+        goto out;
+    }
+
+    rc = snprintf(buf, sizeof(buf), "%lu", info.mi_used_bytes);
+    if (rc < 0) {
+        return EBADMSG;
+    } else if (rc >= sizeof(buf)) {
+        return EMSGSIZE;
+    }
+
+    yaml_scalar_event_initialize(
+        &event,
+        NULL,
+        (yaml_char_t *)YAML_STR_TAG,
+        (yaml_char_t *)buf,
+        strnlen(buf, sizeof(buf)),
+        1,
+        0,
+        YAML_PLAIN_SCALAR_STYLE);
+    if (!yaml_emitter_emit(&emitter, &event)) {
+        rc = EIO;
+        goto out;
+    }
+
+    yaml_mapping_end_event_initialize(&event);
+    if (!yaml_emitter_emit(&emitter, &event)) {
+        rc = EIO;
+        goto out;
+    }
+
+    yaml_document_end_event_initialize(&event, 0);
+    if (!yaml_emitter_emit(&emitter, &event)) {
+        fprintf(stderr, "Failed to emit event %d: %s\n", event.type, emitter.problem);
+        rc = EIO;
+        goto out;
+    }
+
+    yaml_stream_end_event_initialize(&event);
+    if (!yaml_emitter_emit(&emitter, &event)) {
+        rc = EIO;
+        goto out;
+    }
+
+out:
+    if (emitter.opened)
+        yaml_emitter_delete(&emitter);
```

Comment:
emitter.opened is false, so this is leaking.  Maybe move the call to yaml_emitter_initialize top of function, return ENOMEM directly if it fails, and call yaml_emitter_delete here unconditionally.

Reply:
Above, I zeroed out the emitter struct. All should be good as long as I remove the if check.